### ISSUE TITLE
firebase.config().fetch() returns error "Error: fetch() operation cannot be completed successfully." 

### ISSUE THREAD

**Issue by cod2017** [invertase/react-native-firebase#1376]
### **Issue**  
Followed the below example using Remote Config. But I am getting error. 
```
if (__DEV__) {
  firebase.config().enableDeveloperMode();
}

// Set default values
firebase.config().setDefaults({
  hasExperimentalFeature: false,
});

firebase.config().fetch()
  .then(() => {
    return firebase.config().activateFetched();
  })
  .then((activated) => {
    if (!activated) console.log('Fetched data not activated');
    return firebase.config().getValue('hasExperimentalFeature');
  })
  .then((snapshot) => {
    const hasExperimentalFeature = snapshot.val();

    if(hasExperimentalFeature) {
      // call some function
    }

    // continue booting app
  })
  .catch(console.error);
```

It returns an error : Error: fetch() operation cannot be completed successfully

![screen shot 2018-08-07 at 3 25 50 pm](https://user-images.githubusercontent.com/26476946/43769562-a1df0a0c-9a57-11e8-9477-8741e331695b.png)

### Environment

1. Application Target Platform: Android
2. Development Operating System: MacOS High Sierra
3. Build Tools: Xcode 
   buildToolsVersion = "27.0.3"
    minSdkVersion = 16
    compileSdkVersion = 27
    targetSdkVersion = 27
    supportLibVersion = "27.1.0"
4. `React Native` version: 0.56.0
5. `React Native Firebase` Version:  react-native-firebase@4.3.8
6. `Firebase` Module: Remote Config
7. Are you using `typescript`? no

**Comment by Ehesp** [comment]
Hello - the remote config SDKs throw back errors which are less than helpful at times. If you can, open up Android Studio and run your app in it's debug mode, and place a debug marker on [this](https://github.com/invertase/react-native-firebase/blob/master/android/src/main/java/io/invertase/firebase/config/RNFirebaseRemoteConfig.java#L149) line. 

Or, check the logcat output, the internal SDK may log out something which will help.

**Comment by burakduganci** [comment]
I have the same problem. How can we find a solution?

**Comment by Ehesp** [comment]
@burakduganci please see my above comment.

**Comment by ChiAnTai** [comment]
I have some problem that the fetch promise give me no response, and it is fixed when I upgrade simulator to API 28.

**Comment by Ehesp** [comment]
Please respond with some debugging information and we'll reopen. Thanks.

**Comment by ChiAnTai** [comment]
@Ehesp I removed it since API 28 is too new, but I believe the error should be "Google Play services out of date. Requires 12451000 but found 12221470".(from my Google search record)

**Comment by rohit75** [comment]
facing the same issue. Any solution?

**Comment by dembeEdward** [comment]
Any solution yet?

**Comment by mikehardy** [comment]
@dembeEdward this is a closed issue with no debugging information, if you're having issues please open a new issue and follow the template so we know all the versions and steps to reproduce. Also, highly recommend you try to reproduce it either from react-native-firebase-starter or https://github.com/mikehardy/rnfbdemo so you make sure it is not something wrong in your project

**Comment by jezreel8858** [comment]
My solution was to put the fetch function inside a Promise.
Ex:
try {
  return new Promise(
    (resolve) => {
      resolve(
      firebase.config()
      fetch()
        .then(() => {
          return firebase.config().activateFetched();
        })
        .then((activated) => {
          if (!activated) console.log('Fetched data not activated');
          return firebase.config().getValue('hasExperimentalFeature');
        })
        .then((snapshot) => {
          const hasExperimentalFeature = snapshot.val();

          if (hasExperimentalFeature) {
            // call some function
          }

          // continue booting app
        })
        .catch(console.error)
    )
  );
} catch (err) {
  console.warn(err)
}

**Comment by OtacilioN** [comment]
I am facing the same issue =/

**Comment by OtacilioN** [comment]
I am facing the same issue, but adding in a promise did not work @jezreel8858 =/ 

Does anyone have a solution for this?

**Comment by manakuro** [comment]
I followed what @Ehesp said (https://github.com/invertase/react-native-firebase/issues/1376#issuecomment-411031765) and got expired error in Android Studio like this.

```
E/FirebaseRemoteConfig: Fetch failed! Server responded with an error.
    com.google.android.gms.internal.firebase_remote_config.zze: 400 Bad Request
    {
      "code": 400,
      "errors": [
        {
          "domain": "global",
          "message": "API key expired. Please renew the API key.",
          "reason": "badRequest"
        }
      ],
      "message": "API key expired. Please renew the API key.",
      "status": "INVALID_ARGUMENT"
    }
```

That might be one of the reasons of this issue.

**Comment by radiodario** [comment]
This is also failing intermittently for our team, which means that some features are unavailable to our users - Is this related to firebase or something to do with this library?

**Comment by mikehardy** [comment]
Are you able to fetch logcat when it happens? Likely something there indicating the reason. If I recall correctly the throttling on remote config happens much faster than I would have though, maybe that's the underlying cause, as a guess

**Comment by Alaa-Ben** [comment]
Hey guys ! In my case, I had the same error message and logcat didn't show anything until i made it talk :p 
Go to ReactNativeFirebaseConfigModule.java, find the fetch method and do the following:
```
public void fetch(double expirationDurationSeconds, Promise promise) {
    module.fetch((long) expirationDurationSeconds).addOnCompleteListener(task -> {
      if (task.isSuccessful()) {
        promise.resolve(resultWithConstants(task.getResult()));
      } else {
        // add this line
        Log.e("Tag", "ErrorMessage", task.getException());
        rejectPromiseWithConfigException(promise, task.getException());
      }
    });
  }
```

You'll now see the right error message. In my case, it was an authentication error (unauthorized).

**Comment by azizhk** [comment]
I patched `RNFirebaseRemoteConfig.java` to see what the underlying message was.
One of the errors I'm getting is "The client had an error while calling the backend".
The only other instances of this error were: 
- https://github.com/firebase/firebase-android-sdk/issues/1111
- https://github.com/firebase/firebase-android-sdk/issues/1515
- https://stackoverflow.com/questions/60595972/firebaseremoteconfigclientexception-on-production-android

Though could not find anything conclusive, they state that either there was an error with Network, or SSL Handshake.

Here is my patch btw:
```diff
--- a/node_modules/react-native-firebase/android/src/main/java/io/invertase/firebase/config/RNFirebaseRemoteConfig.java
+++ b/node_modules/react-native-firebase/android/src/main/java/io/invertase/firebase/config/RNFirebaseRemoteConfig.java
@@ -154,9 +154,14 @@ class RNFirebaseRemoteConfig extends ReactContextBaseJavaModule {
                 task.getException()
               );
             } else {
+              Exception exception = task.getException();
+              String message = "";
+              if (exception != null) {
+                message = exception.getMessage();
+              }
               promise.reject(
                 "config/failure",
-                "fetch() operation cannot be completed successfully.",
+                "fetch() operation cannot be completed successfully." + message,
                 task.getException()
               );
             }
```

**Comment by gvitelli** [comment]
Hi everyone. @azizhk thanks for the patching snippet to unhide the underlying problems behind the `fetch()` call failure.
Should a new issue be created as a followup to enhance the transparency on these cases? Or are we deeming this as a non-issue?

**Comment by mikehardy** [comment]
If this one is closed then perhaps - also please :pray: check the upstream android / ios SDK release notes as one of them (I forget which sorry) had a mention of something that seemed like it was in the area

**Comment by Paduado** [comment]
I'm having this happening in production Android Apps, I'm logging the `nativeErrorMessage` prop of the error in crashlytics and I'm seeing 2 different messages:
- "The client had an error while calling the backend!"
- "Failed to get Firebase Instance ID token for fetch."

According to [this comment](https://github.com/firebase/firebase-android-sdk/issues/1515#issuecomment-635440621), a full stack trace is needed to correctly identify the issue for the first error, any suggestions on how could I log that without patching the native package?

Also suggestions for debugging the second error would be greatly appreciated 🙏

**Comment by mikehardy** [comment]
@Paduado patch the native code with wild abandon, it is very easy. You just go into node_modules and hack in exactly what you need, then use https://github.com/ds300/patch-package to make it persistent (that's the key trick).

**Comment by Paduado** [comment]
So I made the patch @mikehardy and I'm finally seeing the native stack on the crashlytics console, this is the error that I'm seeing more frequently, any ideas on what could be the issue?
```
App: RemoteConfig native error message: com.google.firebase.remoteconfig.g: The client had an error while calling the backend! 
at com.google.firebase.remoteconfig.internal.ConfigFetchHttpClient.fetch() 
at com.google.firebase.remoteconfig.internal.k.a() 
at com.google.firebase.remoteconfig.internal.k.b() 
at com.google.firebase.remoteconfig.internal.k.a() 
at com.google.firebase.remoteconfig.internal.h.a() 
at d.d.b.d.h.o.run() 
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113) 
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588) 
at java.lang.Thread.run(Thread.java:833) Caused by: java.net.UnknownHostException: Unable to resolve host "firebaseremoteconfig.googleapis.com": No address associated with hostname 
at java.net.InetAddress.lookupHostByName(InetAddress.java:470) 
at java.net.InetAddress.getAllByNameImpl(InetAddress.java:252) 
at java.net.InetAddress.getAllByName(InetAddress.java:215) 
at com.android.okhttp.internal.Network$1.resolveInetAddresses(Network.java:29) 
at com.android.okhttp.internal.http.RouteSelector.resetNextInetSocketAddress(RouteSelector.java:188) 
at com.android.okhttp.internal.http.RouteSelector.nextProxy(RouteSelector.java:157) 
at com.android.okhttp.internal.http.RouteSelector.next(RouteSelector.java:100) 
at com.android.okhttp.internal.http.HttpEngine.createNextConnection(HttpEngine.java:368) 
at com.android.okhttp.internal.http.HttpEngine.nextConnection(HttpEngine.java:351) 
at com.android.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:341) 
at com.android.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:259) 
at com.android.okhttp.internal.huc.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:454) 
at com.android.okhttp.internal.huc.HttpURLConnectionImpl.connect(HttpURLConnectionImpl.java:114) 
at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getOutputStream(HttpURLConnectionImpl.java:245) 
at com.android.okhttp.internal.huc.DelegatingHttpsURLConnection.getOutputStream(DelegatingHttpsURLConnection.java:218) 
at com.android.okhttp.internal.huc.HttpsURLConnectionImpl.getOutputStream(HttpsURLConnectionImpl.java:25) 
at com.google.firebase.remoteconfig.internal.ConfigFetchHttpClient.a() ... 9 more Caused by: android.system.GaiException: android_getaddrinfo failed: EAI_NODATA (No address associated with hostname) 
at libcore.io.Posix.android_getaddrinfo(Native Method) 
at libcore.io.ForwardingOs.android_getaddrinfo(ForwardingOs.java:55) 
at java.net.InetAddress.lookupHostByName(InetAddress.java:451) ... 25 more
```

**Comment by mikehardy** [comment]
Transient DNS failure? From your stack

> at java.lang.Thread.run(Thread.java:833) Caused by: java.net.UnknownHostException: Unable to resolve host "firebaseremoteconfig.googleapis.com": No address associated with hostname

**Comment by mikehardy** [comment]
So upstream others saw similar and the closing comment in particular is the most important: https://github.com/firebase/firebase-android-sdk/issues/1069#issuecomment-599212637

So for this to be actionable here , the error would have to actually escape and crash the app - otherwise it's a "typical" sort of network error and would happen any time the phone was in airplane mode (as mentioned in that comment)

The original issue here seems to be that it was unexplained, not that it was crashing, so I don't think there's a crash fix to implement or anything.

It's possible that this could be added to the tips-n-tricks section https://rnfirebase.io/faqs-and-tips or that the API itself could have a comment added with an edit here https://github.com/invertase/react-native-firebase/blob/525dcc29de36c66d64dcbb404dfeb6de17392a11/packages/remote-config/lib/index.d.ts#L396-L408 - not sure what is more useful for a user, probably adding the comment to fetch "this is a network request and will fail if the mobile network is not functioning when the method is called" ?

**Comment by Paduado** [comment]
Yeah that makes sense, thanks a lot for your help!

**Comment by lustig-bakkt** [comment]
For any idiots out there like me, my emulator wasn't connected to the Internet. 

This StackOverflow post fixed it: https://stackoverflow.com/questions/42736038/android-emulator-not-able-to-access-the-internet

**TL;DR** change your laptop's DNS to 8.8.8.8

**Comment by rajeevverma076** [comment]
NativeFirebaseError: [remoteConfig/failure] fetch() operation cannot be completed successfully. at FirebaseConfigModule.fetchAndActivate
